libcudacxx/include/cuda/std/__string/helper_functions.h

```diff
@@ -0,0 +1,294 @@
+    {
+      return __last1;
+    }
+
+    // It is faster to compare from the first byte of __first1 even if we
+    // already know that it matches the first byte of __first2: this is because
+    // __first2 is most likely aligned, as it is user's "pattern" string, and
+    // __first1 + 1 is most likely not aligned, as the match is in the middle of
+    // the string.
+    if (_Traits::compare(__first1, __first2, __len2) == 0)
+    {
+      return __first1;
+    }
+
+    ++__first1;
+  }
+}
+
+template <class _CharT, class _SizeT, class _Traits, _SizeT __npos>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _SizeT
+__cccl_str_find(const _CharT* __p, _SizeT __sz, const _CharT* __s, _SizeT __pos, _SizeT __n) noexcept
+{
+  if (__pos > __sz)
+  {
+    return __npos;
+  }
+
+  if (__n == 0) // There is nothing to search, just return __pos.
+  {
+    return __pos;
+  }
+
+  const _CharT* __r = _CUDA_VSTD::__cccl_search_substring<_CharT, _Traits>(__p + __pos, __p + __sz, __s, __s + __n);
+
+  if (__r == __p + __sz)
+  {
+    return __npos;
+  }
+  return static_cast<_SizeT>(__r - __p);
+}
+
+template <class _CharT, class _SizeT, class _Traits, _SizeT __npos>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _SizeT
+__cccl_str_rfind(const _CharT* __p, _SizeT __sz, _CharT __c, _SizeT __pos) noexcept
+{
+  if (__sz < 1)
+  {
+    return __npos;
+  }
+  if (__pos < __sz)
+  {
+    ++__pos;
+  }
+  else
+  {
+    __pos = __sz;
+  }
+  for (const _CharT* __ps = __p + __pos; __ps != __p;)
+  {
+    if (_Traits::eq(*--__ps, __c))
+    {
+      return static_cast<_SizeT>(__ps - __p);
+    }
+  }
+  return __npos;
+}
+
+template <class _CharT, class _SizeT, class _Traits, _SizeT __npos>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _SizeT
+__cccl_str_rfind(const _CharT* __p, _SizeT __sz, const _CharT* __s, _SizeT __pos, _SizeT __n) noexcept
+{
+  __pos = _CUDA_VSTD::min(__pos, __sz);
+  if (__n < __sz - __pos)
+  {
+    __pos += __n;
+  }
+  else
+  {
+    __pos = __sz;
+  }
+  const _CharT* __r = _CUDA_VSTD::__find_end(
+    __p, __p + __pos, __s, __s + __n, _Traits::eq, random_access_iterator_tag(), random_access_iterator_tag());
+  if (__n > 0 && __r == __p + __pos)
+  {
+    return __npos;
+  }
+  return static_cast<_SizeT>(__r - __p);
+}
+
+template <class _CharT, class _SizeT, class _Traits, _SizeT __npos>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _SizeT
+__cccl_str_find_first_of(const _CharT* __p, _SizeT __sz, const _CharT* __s, _SizeT __pos, _SizeT __n) noexcept
+{
+  if (__pos >= __sz || __n == 0)
+  {
+    return __npos;
+  }
+  const _CharT* __r = _CUDA_VSTD::__find_first_of_ce(__p + __pos, __p + __sz, __s, __s + __n, _Traits::eq);
+  if (__r == __p + __sz)
+  {
+    return __npos;
+  }
+  return static_cast<_SizeT>(__r - __p);
+}
+
+template <class _CharT, class _SizeT, class _Traits, _SizeT __npos>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _SizeT
+__cccl_str_find_last_of(const _CharT* __p, _SizeT __sz, const _CharT* __s, _SizeT __pos, _SizeT __n) noexcept
+{
+  if (__n != 0)
```

Comment:
Style:  I strongly prefer early returns whenever possible, so this could just be:
```suggestion
  if (__n == 0) {
    return __npos;
  }
```